### DISCUSSION TITLE
Connecting the website component in a seamless way

### DISCUSSION THREAD

**Discussion by tsuf239** [winglang/wing#3574] [Ideas]
I was thinking about how to make Wing more appealing to website developers, keeping in mind that the connectivity between the two should be seamless, and work both on sim and the cloud.
I came out with this interesting approach, and I would love to hear what you think about it.
I'll start with the end: 
* As a full stack developer, to use my backend components, I use HTTP calls. It's a better practice to wrap those in a "wrapper" service, that also takes care of auth and error handling, and is usually built once. when a developer wants to add a call or use an existing one, they either call the wrapper function with the route path and request payload, sometimes the state-management framework is taking care of a part of this workflow, so the call results in a single value change in the store.
* The data is usually transferred via HTTP calls, sometimes using other protocols, or using some sort of pub/sub (won't be discussed in this discussion, but can be added in the future)

what if:
1. The wing file/files will be located next to the website source files, or pointed to them in any way
2. we'll introduce the `expose` keyword to expose a function to the website js files:
```
expose("function name in js files", inflight handler, props : {path, method, auth, errorHandling (can be written in wing or an "externed" js) });
```
for example: 
```
let api = new cloud.Api();
let counter = new cloud.Counter();

let increase = inflight () => {
  return { body:  counter.inc() + 1 };
};

let peek = inflight () => {
  return { body:  counter.peek() };
};

api.post("/counter", increase);
api.get("/counter", peek);

//Then some really heavy error prone lifting to make this appear in the website js files
```
will be equivalent of:
```
let counter = new cloud.Counter();

let increase = inflight () => {
  return { body:  counter.inc() + 1 };
};

let peek = inflight () => {
  return { body:  counter.peek() };
};

//This line will start the magic
// this will actually creates both api gateway, and HTTP call to the gw, and exports it to a outdir of a choice, in the future we can export types as well

expose("increaseCounter", increase, {method: Http.POST, path: "/counter"}); 
expose("peekCounter", peek, {method: Http.GET, path: "/counter"});

```
other option is:
```
let api = new cloud.Api();
let counter = new cloud.Counter();

let increase = inflight () => {
  return { body:  counter.inc() + 1 };
};

let peek = inflight () => {
  return { body:  counter.peek() };
};

let incRoute = api.post("/counter", increase);
let peekRoute = api.get("/counter", peek);

expose("increaseCounter", incRoute, {errorHandling and stuff});
expose("peekCounter", peekRoute, {errorHandling and stuff});
```
(this way if we want to supply domain or to tweak an existing API instance, we definitely can)

3. Then the wing file/s (that is specified in any config file) is compiled along the regular webpack/babel compilation of the UI framework (let's say react)
The "npm start" command starts the simulator, and "npm build/serve" will deploy to the chosen target (specified in the configuration file as well)
4. an API gateway is created with all the "exposed" methods, URL and type of environment (dev, prod, stage, test, etc... if needed) can be set up ahead of time or updated on the fly
5. a js file is added to the website src files along with the compilation/hot reloading of the framework, which already contains all "exposed" methods, in a beautifully organized well-maintained, error-handled "fetch/http" service, ready to use in a state management of your choice
7. Let the developer approach the "exposed" method as easily as:

```JS
import React, { useState, useEffect } from 'react';
import { increaseCounter, peekCounter } from "../wing" // THAT'S THE MAGIC!!!

function Example() {
  const [count, setCount] = useState(0);

 const increase = async () => {
    // this wing-exposed fn is already error proof 🤩 
     const res  =  await increaseCounter();
     setCount(res.body);
  }

const peek = async () => {
    // this wing-exposed fn is already error proof 🤩 
     const res  =  await peekCounter();
     setCount(res.body);
  }

  useEffect(() => {
// will initiate the counter value on load
   peek();
  }, []);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={increase}>
        Click me
      </button>
    </div>
  );
}
```

**Comment by tsuf239** [comment]
I just thought of it, we can also integrate in the same way to a backend already written in nodeJS (as well as other JSII compatible languaguges!)

**Comment by skyrpex** [comment]
I love that we're already talking about this stuff. We need this!

Here's some feedback:

- The `expose` API (pun intended) seems to be a singleton. That may be a problem in big projects, or in a microservices-like project
- Providing a working implementation of a client may be handy but I'd appreciate that the generated file allowed to bring your custom implementation

So, what if:

- Instead of exposing HTTP endpoints, we allow exposing resources
- What gets exposed isn't a fully working client but the necessary data to create a client

For example:

```wing
let api = new cloud.Api();
let counter = new cloud.Counter();

let increase = inflight () => {
  return { body:  counter.inc() + 1 };
};

let peek = inflight () => {
  return { body:  counter.peek() };
};

let incRoute = api.post("/counter", increase);
let peekRoute = api.get("/counter", peek);

expose("myApi", api);
```

```ts
import React, { useState, useEffect } from 'react';
import { createApiClient } from "@winglang/client"; // Our standard implementation. Can use your own
import { myApi } from "../wing"; // Generated file, do not touch. Includes URL, routes and their types

const { increaseCounter, peekCounter, peek } = createApiClient(myApi); // Typed client

function Example() {
  const [count, setCount] = useState(0);

 const increase = async () => {
    // this wing-exposed fn is already error proof 🤩 
     const res  =  await increaseCounter();
     setCount(res.body);
  }

const peek = async () => {
    // this wing-exposed fn is already error proof 🤩 
     const res  =  await peekCounter();
     setCount(res.body);
  }

  useEffect(() => {
// will initiate the counter value on load
   myApi.peek();
  }, []);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={increase}>
        Click me
      </button>
    </div>
  );
}
```

**Comment by eladb** [comment]
One way to think about this is that web apps are not very different from inflight code. The fact that they run on client browser computer and not in a cloud compute platform doesn't change the inflight/preflight nature of this relationship.

We've actually prototyped stuff like this in one of the offsite where we implemented a `cloud.Webapp` resource which could execute inflight closures in the client.

**Reply by bitfrost** [reply]
Perhaps you could add 'landed/local' or 'post-flight' for client targets.  I.e. <browser|ios|android> as a deeper compute target :)    

The bridge to the humans/clients seems to be the missing piece in the wing model. A lot of energy is currently being poured into inflight/local in a similar way in other frameworks.   Where wing bridges preflight/inflight React.js is pushing into inflight/local (i.e. React Server Components / Next.js ) Both have very very similar DX improvements.  I am greedy and want all 3.   preflight/inflight/local

**Comment by ekeren** [comment]
Maybe we should consider adding another resource `js.SDK` for this kind of example

```ts
bring cloud;
bring http;

let api = new cloud.Api();
let counter = new cloud.Counter();

let increase = inflight (): Json => {
  return { body:  counter.inc() + 1 };
};

let peek = inflight (): Json => {
  return { body:  counter.peek() };
};


// public API
api.post("/counter", increase);
api.get("/counter", peek);

let sdk = new js.SDK();

// this SDK relies on our public API 
sdk.addFunction("increase", inflight ():http.Response => {
  return http.get(api.url + "/counter");
});
// this SDK call relies on a private API (there is no API)
sdk.addFunction("peek", inflight ():http.Response => {
  return peek();
});
```
